packages/common/src/hooks/useHarmonyField.ts

```diff
@@ -0,0 +1,262 @@
+   * field. Do not use this for eg. debouncing validation API requests unless y
+   * you use it for _every_ field of the form (or other fields don't validate
+   * on change). Otherwise changes to other fields will trigger non-debounced
+   * validation requests anyway.
+   *
+   * Requires validateOnChange to be true (either for the form or this field).
+   *
+   * The default of 0 means errors will be hidden until the user unfocuses.
+   *
+   * @default 0
+   */
+  debouncedIdleMs?: number
+  /**
+   * Whether to show errors prior to the first submission attempt.
+   * @default false
+   */
+  showErrorBeforeSubmit?: boolean
+  /**
+   * Whether to validate this particular field on change or not.
+   * If present, overrides the form's validateFieldOnChange.
+   * @default undefined
+   */
+  validateFieldOnChange?: boolean
+}
+
+/**
+ * Wrapper around Formik's {@link useFormikField useField} that modifies the
+ * returned results for Harmony compatible behavior, including the following
+ * configurable behaviors:
+ * - Only show errors after a submission has been attempted
+ * - Apply transformations to the value on blur or change
+ * - Debounce validations and error displays to lighten the validation
+ *   workload or hide the errors while the user is still typing.
+ * - Allow overriding the Formik context's "validateOnChange"
+ *
+ * Most of the extra functionality is reserved for fields that users type into,
+ * eg. text, email, password inputs.
+ * Unlike the useField hook from Formik, this implementation relies much less
+ * on the event target in onChange and onBlur, instead using the passed in
+ * field name path. This makes usage with ReactNative more straightforward.
+ *
+ * @param propsOrFieldName the field's name, or the full props for the hook.
+ * @example
+ * const MyComponent = () => {
+ *    const [emailField, { error }] = useHarmonyField({ name: 'email', label: 'Email' })
+ *    return <TextInput {...field} error={!!error} helperText={error} />
+ * }
+ */
+export const useHarmonyField = <Value = any>(
+  propsOrFieldName: string | UseHarmonyFieldProps<Value>
+): [
+  FormikFieldInputProps<Value>,
+  FieldMetaProps<Value>,
+  FieldHelperProps<Value>
+] => {
+  const props =
+    typeof propsOrFieldName === 'string'
+      ? { name: propsOrFieldName }
+      : propsOrFieldName
+  const {
+    name,
+    type = 'text',
+    debouncedIdleMs = 0,
+    transformValueOnChange,
+    transformValueOnBlur,
+    showErrorBeforeSubmit = false,
+    validateFieldOnChange
+  } = props
+  const [field, meta, helpers] = useFormikField(props)
+  const { touched, error } = meta
+  const { setValue, setTouched } = helpers
+  const {
+    validateField,
+    validateOnBlur,
+    validateOnChange: validateFormOnChange,
+    submitCount
+  } = useFormikContext()
+
+  // Tracks whether the user is actively editing the field
+  const [isChanging, setIsChanging] = useState(false)
```

Comment:
love this. It baffles me that Formik didn't really come with this concept built in 💀

Reply:
indeed!